### PR TITLE
Inject trace ids into paperplane logs

### PR DESCRIPTION
**Pull request by flintinatux**
### What does this PR do?

Adds support for log injection in `paperplane`.

### Motivation

When I first made the plugin for `paperplane`, I followed the examples of `express` and `hapi`.  Those don't have built-in logging, but `paperplane` does, and I totally overlooked it.  We'd like our logs to be correlated with traces in APM, so this PR aims to enable that.

### Plugin Checklist

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [x] API [documentation][3].
- [x] CircleCI [jobs/workflows][4].
- [x] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/src/plugins/index.js

### Additional Notes

This does bump the supported version of `paperplane` to `v2.3.1`.  That's because log injection on error logs doesn't work before that version.

### CODE REVIEW

File: packages/datadog-plugin-paperplane/src/index.js [hunk 1]
```diff
@@ -36,17 +55,22 @@ const wrapRoutes = tracer => routes => handlers => {
...
 
 function unpatch (paperplane) {
-  this.unwrap(paperplane, ['mount', 'routes'])
+  this.unwrap(paperplane, ['logger', 'mount', 'routes'])
 }
 
 module.exports = {
   name: 'paperplane',
-  versions: ['>=2.3'],
+  versions: ['>=2.3.1'],
```
Comment (rochdev) [inline packages/datadog-plugin-paperplane/src/index.js]: Why the version bump?

Comment (flintinatux) [reply packages/datadog-plugin-paperplane/src/index.js]: Because this was needed to make log injection work for errors:
https://github.com/articulate/paperplane/pull/46

**Comment by flintinatux** [comment]
@rochdev, thanks for the heads up on the linting.  I ran `yarn lint` when I was spiking out the code, but forgot to do it again after I finished up the tests.  😅

File: packages/datadog-plugin-paperplane/src/index.js (hunk 1 above)
Comment (rochdev) [reply packages/datadog-plugin-paperplane/src/index.js]: Support for 2.3.0 could be kept but would just be missing log injection.

Comment (flintinatux) [reply packages/datadog-plugin-paperplane/src/index.js]: You know, that's a good point.  Let me see if I can make that work (probably on Monday).  When I do, should I document that difference somewhere?

Comment (rochdev) [reply packages/datadog-plugin-paperplane/src/index.js]: > Let me see if I can make that work

You can find a good example of different feature support by version in the [hapi](https://github.com/DataDog/dd-trace-js/blob/403154a57edcc3168668d38a855f05d23c9d0348/packages/datadog-plugin-hapi/src/index.js#L51) plugin.

> When I do, should I document that difference somewhere?

I would say to add a note in the [docs](https://github.com/DataDog/documentation/blob/master/content/en/tracing/setup/nodejs.md#compatibility), or simply add `paperplane` to the list of supported loggers with `>=2.3.1` as the version.

**Comment by flintinatux** [comment]
@rochdev, we realized that patching the `logger` properly required an [internal refactor](https://github.com/articulate/paperplane/pull/47) to make monkey-patching with `shimmer` work correctly.  Otherwise, exported functions like `logger` that are [reused by other functions](https://github.com/articulate/paperplane/blob/master/lib/mount.js#L9) wouldn't be patched internally.  Consequently, we've bumped `paperplane` yet again to `v2.3.2`.

However, I was able to continue support for `v2.3.0-2.3.1`, just without log injection, as you suggested.  Please let me know what you think.  Thanks!

File: packages/datadog-plugin-paperplane/src/index.js [hunk 2]
```diff
@@ -53,24 +53,53 @@ const wrapRoutes = tracer => routes => handlers => {
...
+    patch (exports, tracer) {
+      this.wrap(exports, 'routes', wrapRoutes(tracer))
+    },
+    unpatch (exports) {
+      this.unwrap(exports, 'routes')
+    }
+  },
+  {
+    name: 'paperplane',
+    versions: ['2.3.0 - 2.3.1'],
```
Comment (rochdev) [inline packages/datadog-plugin-paperplane/src/index.js]: It looks like both `lib/routes.js` and `lib/mount.js` were available in these versions as well. Could this block simply be removed and the requirements above reduced?

File: packages/datadog-plugin-paperplane/src/index.js [hunk 3]
```diff
@@ -53,24 +53,53 @@ const wrapRoutes = tracer => routes => handlers => {
...
 
-function patch (paperplane, tracer, config) {
-  config = web.normalizeConfig(config)
-
-  if (tracer._logInjection) {
-    this.wrap(paperplane, 'logger', wrapLogger(tracer))
+module.exports = [
+  {
+    name: 'paperplane',
+    versions: ['>=2.3.2'],
```
Comment (rochdev) [inline packages/datadog-plugin-paperplane/src/index.js]: Just out of curiosity, why the change from `2.3.1` to `2.3.2`?

Comment (rochdev) [reply packages/datadog-plugin-paperplane/src/index.js]: Just saw your answer above 👍

File: packages/datadog-plugin-paperplane/src/index.js (hunk 2 above)
Comment (flintinatux) [reply packages/datadog-plugin-paperplane/src/index.js]: I originally thought I'd do that, but it won't work.  The files themselves were available then, but internally they exported as the default export, rather than a named export.  For example:

**`lib/routes.js` in 2.3.1:**
```js
module.exports = curry(routes)
```

**`lib/routes.js` in 2.3.2:**
```js
exports.routes = curry(routes)
```

The named exports in 2.3.2 was done to support internal patching, and it's not supported prior to that.  So I had to patch them differently for the earlier versions.

Comment (flintinatux) [reply packages/datadog-plugin-paperplane/src/index.js]: We went ahead and did named exports for all internal functions, rather than just `logger`, to avoid any more future bumps to support patching in Datadog or any other third-party library.

Comment (rochdev) [reply packages/datadog-plugin-paperplane/src/index.js]: But then doesn't the opposite apply? (using the named export from the library itself instead of the internal files).

It's also possible to overwrite the default export using the instrumenter, but it comes with a few limitations so the additional block is probably better to avoid that.

I'm just trying to reduce the number of patches if possible 😄

Comment (flintinatux) [reply packages/datadog-plugin-paperplane/src/index.js]: If we patch it internally, then the externally exported reference is also patched.  `shimmer` and `require-in-the-middle` do a good job of allowing that.

And just to be sure:  the patch is done internally for `v2.3.2`, but in the tests [I'm still testing the externally exported functions](https://github.com/DataDog/dd-trace-js/pull/633/files#diff-b5639a388194ae9b8716861596960d06R532), and it works just fine.

Comment (rochdev) [reply packages/datadog-plugin-paperplane/src/index.js]: What I meant was, why not just always patch externally as it was before? In general that is the preferred method so that the instrumentation continues to work even if the internal structure changes. Instrumenting internal modules is a last resort for when what we need to patch is not exported.

Comment (flintinatux) [reply packages/datadog-plugin-paperplane/src/index.js]: I totally understand.  But these are first-class functions we are patching/replacing, not prototype methods.  So if I patched just the external export, it would just be patching `require('paperplane').logger`.  Internally, when `lib/mount.js` uses the logger, it would not be patched, because it uses `const { logger: defaultLogger } = require('./logger')`.

I get it.  I would rather patch just the interface.  But in this case, if I just patch externally, and I don't supply a custom logger, it would use the default, and no logs would be injected.

Comment (rochdev) [reply packages/datadog-plugin-paperplane/src/index.js]: Fair enough.

**Comment by rochdev** [comment]
@flintinatux Is it supported to use the logger outside the scope of `paperplane`? For example using it as the logger for the tracer like this: `tracer.init({ logger: paperplane.logger })`.

For that to work, the logger (and thus `paperplane`) needs to be imported before the tracer, which wouldn't work with the current approach if I'm not mistaken.

**Comment by flintinatux** [comment]
That wasn't really an intended use case for us, mostly because I always thought you had to require the `tracer` before anything else that you wanted to be patched.

Are there plugins that can patch after-the-fact like that?

**Comment by flintinatux** [comment]
Oh, but the answer to your first question:  yes, we do use `paperplane.logger` as a general use logger outside of the `paperplane` request cycle.  It helps serialize POJO's and `Error`'s as JSON for parsing with log aggregators like Datadog, etc.

**Comment by rochdev** [comment]
> Are there plugins that can patch after-the-fact like that?

Loggers mostly 😅. It can work in some cases with other plugins as well, but the ones that we explicitly tested were loggers since to use them with the tracer they have to be imported beforehand.

Do you think there's a way this could work with `paperplane` as well? I'll still merge this PR anyway regardless, but if it can be handled, it would definitely be nice to have it. Otherwise users will have to choose between using the built-in logger with the tracer, or having log injection.

**Comment by rochdev** [comment]
I wonder if some kind of check could be added if the logger was already patched, and patch it otherwise, in both the public API and the internal one. This would technically result in 2 different logger instances when patched after the fact, but in general that wouldn't cause issues.

@flintinatux Do you think that would work? Another option could be to proxy the call in the index, so that it's just delegating to the underlying logger 🤔 Something like `exports.logger = record => logger.logger(record)`.

**Comment by flintinatux** [comment]
If I had to guess, it works for other loggers because they are OOP, so they can patch the `prototype` after-the-fact, and it swaps out the underlying implementation, even for existing instances.

That won't work for `paperplane` since it's 💯 FP.  All of its functions are first class, not prototype methods.  And since functions are passed by reference, swapping out the export after-the-fact wouldn't affect existing references.

However, I think you could work around it with the magic of javascript variable scoping:
```js
// wrap it as a custom logger
const myLogger = x =>
  paperplane.logger(x)

// init with your custom logger
const tracer = require('dd-trace').init({ logger: myLogger })

// then require paperplane after it's been patched
const paperplane = require('paperplane')
```

**Comment by rochdev** [comment]
Makes sense. This won't work for anything logged while the tracer is initializing, but I think that's a reasonable limitation. We could always solve this in the tracer by deferring the actual call to the logger to the next tick.
